integrations/amazon_bedrock/src/haystack_integrations/components/downloaders/s3/s3_downloader.py

```diff
@@ -0,0 +1,405 @@
+
+        def resolve_secret(secret: Optional[Secret]) -> Optional[str]:
+            return secret.resolve_value() if secret else None
+
+        try:
+            session = get_aws_session(
+                aws_access_key_id=resolve_secret(aws_access_key_id),
+                aws_secret_access_key=resolve_secret(aws_secret_access_key),
+                aws_session_token=resolve_secret(aws_session_token),
+                aws_region_name=resolve_secret(aws_region_name),
+                aws_profile_name=resolve_secret(aws_profile_name),
+            )
+            config = Config(
+                user_agent_extra="x-client-framework:haystack", **(self.boto3_config if self.boto3_config else {})
+            )
+
+            self._client = session.client("s3", config=config)
+        except Exception as exception:
+            msg = (
+                "Could not connect to Amazon S3. Make sure the AWS environment is configured correctly. "
+                "See https://boto3.amazonaws.com/v1/documentation/api/latest/guide/quickstart.html#configuration"
+            )
+            raise AmazonS3ConfigurationError(msg) from exception
+
+    def warm_up(self) -> None:
+        """Ensure local cache directory exists."""
+        self.file_root_path.mkdir(parents=True, exist_ok=True)
+
+    @component.output_types(
+        documents=List[Document],
+        sources=List[Union[str, Path, ByteStream]],
+    )
+    def run(
+        self,
+        documents: Optional[List[Document]] = None,
+        sources: Optional[List[Union[ByteStream, UUID, str]]] = None,
+    ) -> Dict[str, Union[List[Document], List[Union[str, Path, ByteStream]]]]:
+        """Download S3-linked files and return enriched `Document`s and source paths.
+
+        You can pass existing `Document`s (with S3 coordinates in `meta`) and/or `sources`
+        (e.g., `s3://...` URLs or `ByteStream` with S3 metadata). Files are cached locally.
+
+        :param documents: Documents to download. Each must include one of:
+            - `meta['s3_bucket']` + `meta['s3_key']`, or
+            - `meta['s3_url']` (e.g., `s3://bucket/key`), or
+            - `meta['s3_key']` (with default bucket provided elsewhere).
+            Non-conforming documents are skipped.
+        :param sources: Optional heterogeneous inputs used to build `Document`s:
+            - `str` with `s3://bucket/key`,
+            - `ByteStream` whose `meta` contains `s3_url` or (`s3_bucket` + `s3_key`),
+            - Other types are ignored with a warning.
+        :returns: A dictionary with:
+            - `documents`: The downloaded `Document`s; each has `meta['file_path']` and may include `mime_type`.
+            - `sources`: The sources in the requested target type (`str`, `Path`, or `ByteStream`).
+        :raises AmazonS3Error: If a download attempt fails.
+        :raises ValueError: If an S3 URL in inputs is invalid.
+        """
+        # Ensure cache dir
+        self.warm_up()
```

Comment:
Hmm, we never call `warm_up` in the run method. We should instead follow our normal pattern of throwing an error if we detect that the component has not warmed up.